lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+    }
+    return eig_solver;
+  }
+
+  // Utilities and functions common to all Eigensolver instances
+  //------------------------------------------------------------------------------
+
+  void EigenSolver::matVec(const DiracMatrix &mat, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+    if (!tmp1 || !tmp2) {
+      ColorSpinorParam param(in);
+      if (!tmp1) tmp1 = ColorSpinorField::Create(param);
+      if (!tmp2) tmp2 = ColorSpinorField::Create(param);
+    }
+    mat(out, in, *tmp1, *tmp2);
+  }
+
+  void EigenSolver::chebyOp(const DiracMatrix &mat, ColorSpinorField &out, const ColorSpinorField &in)
+  {
+    // Just do a simple matVec if no poly acc is requested
+    if (!eig_param->use_poly_acc) {
+      time_ = -clock();
+      matVec(mat, out, in);
+      time_ += clock();
+      time_mv += time_;
+      return;
+    }
+
+    if (eig_param->poly_deg == 0) { errorQuda("Polynomial acceleration requested with zero polynomial degree"); }
+
+    // Compute the polynomial accelerated operator.
+    double delta, theta;
+    double sigma, sigma1, sigma_old;
+    double d1, d2, d3;
+
+    double a = eig_param->a_min;
+    double b = eig_param->a_max;
+
+    delta = (b - a) / 2.0;
+    theta = (b + a) / 2.0;
+
+    sigma1 = -delta / theta;
+
+    d1 = sigma1 / delta;
+    d2 = 1.0;
+
+    // out = d2 * in + d1 * out
+    // C_1(x) = x
+    time_ = -clock();
+    matVec(mat, out, in);
+    time_ += clock();
+    time_mv += time_;
+
+    time_ = -clock();
+    blas::caxpby(d2, const_cast<ColorSpinorField &>(in), d1, out);
+    if (eig_param->poly_deg == 1) return;
+
+    // C_0 is the current 'in'  vector.
+    // C_1 is the current 'out' vector.
+
+    // Clone 'in' to two temporary vectors.
+    ColorSpinorField *tmp1 = ColorSpinorField::Create(in);
+    ColorSpinorField *tmp2 = ColorSpinorField::Create(in);
+
+    blas::copy(*tmp1, in);
+    blas::copy(*tmp2, out);
+    time_ += clock();
+    time_mb += time_;
+
+    // Using Chebyshev polynomial recursion relation,
+    // C_{m+1}(x) = 2*x*C_{m} - C_{m-1}
+
+    sigma_old = sigma1;
+
+    // construct C_{m+1}(x)
+    for (int i = 2; i < eig_param->poly_deg; i++) {
+
+      sigma = 1.0 / (2.0 / sigma1 - sigma_old);
+
+      d1 = 2.0 * sigma / delta;
+      d2 = -d1 * theta;
+      d3 = -sigma * sigma_old;
+
+      // mat*C_{m}(x)
+      time_ = -clock();
+      matVec(mat, out, *tmp2);
+      time_ += clock();
+      time_mv += time_;
+
+      time_ = -clock();
+      Complex d1c(d1, 0.0);
+      Complex d2c(d2, 0.0);
+      Complex d3c(d3, 0.0);
+      blas::caxpbypczw(d3c, *tmp1, d2c, *tmp2, d1c, out, *tmp1);
+      std::swap(tmp1, tmp2);
+      time_ += clock();
+      time_mb += time_;
+
+      sigma_old = sigma;
+    }
+    blas::copy(out,*tmp2);
+
+    delete tmp1;
+    delete tmp2;
+  }
+
+  // Orthogonalise r against V_[j]
+  Complex EigenSolver::orthogonalize(std::vector<ColorSpinorField *> vecs, std::vector<ColorSpinorField *> rvec, int j)
```

Comment:
Is there any reason we have this around since we have `blockOrthogonalize`?

Reply:
Nope, it has been banished :)